File: src/lib/server/modules/tuttoanimemanga.ts
import type { SearchResult, Manga } from '../classes/interfaces';
import Module from '../classes/Module';
import axios from 'axios';

class TuttoAnimeManga extends Module {
	id = 'tam';
	name = 'TuttoAnimeManga';

	search(query: string): Promise<SearchResult[]> {
		return new Promise(async (resolve: (value: SearchResult[]) => void) => {
			if (query.length < 3 && query !== '') {
				resolve([]);
			}

			const page = await axios.get(
				query === ''
					? 'https://tuttoanimemanga.net/api/comics'
					: 'https://tuttoanimemanga.net/api/search/' + encodeURIComponent(query)
			);

			resolve(
				page.data.comics.map((v: { url: string; title: string }) => ({
					id: v.url,
					title: v.title
				}))
			);
		});
	}

	manga(id: string): Promise<Manga> {
		return new Promise(async (resolve: (value: Manga) => void) => {
			const page = await axios.get('https://tuttoanimemanga.net/api' + id);
			const comic = page.data.comic;

			resolve({
				synopsis: comic.description,
				author: comic.author,
				artist: comic.artist,
				img: comic.thumbnail,
				chapters: comic.chapters.map(
					(v: { full_title: string; url: string; published_on: string }) => ({
						title: v.full_title,
						id: v.url,
						date: new Date(v.published_on)
					})
				),
				sourceurl: 'http://tuttoanimemanga.net' + id
			});
		});
	}
	chapter(manga: string, id: string): Promise<string[]> {
		return new Promise(async (resolve: (value: string[]) => void) => {
			const page = await axios.get('https://tuttoanimemanga.net/api' + id);

			resolve(page.data.chapter.pages);
		});
	}
}

export default new TuttoAnimeManga();

File: pkg/modules/tuttoanimemanga/tuttoanimemanga.go
package tuttoanimemanga

import (
	"github.com/Minettyx/FoolslideProxy/pkg/types"
	"github.com/Minettyx/FoolslideProxy/pkg/utils"
	"net/url"
	"time"
)

var baseurl = "https://tuttoanimemanga.net/"

var TuttoAnimeManga = types.Module{
	Id:    "tam",
	Name:  "TuttoAnimeManga",
	Flags: types.ModuleFlags{},

	Search: func(query string, language *string) ([]types.SearchResult, error) {
		var data searchRes
		err, _ := utils.GetAndJsonParse(baseurl+"api/search/"+url.QueryEscape(query), &data)

		results := []types.SearchResult{}

		for _, com := range data.Comics {
			results = append(results, types.SearchResult{
				Id:    com.Url,
				Title: com.Title,
				Image: com.Thumbnail,
			})

			println(com.Thumbnail)
		}

		return results, err
	},

	Manga: func(id string) (*types.Manga, error) {
		var data mangaRes
		err, _ := utils.GetAndJsonParse(baseurl+"api"+id, &data)
		if err != nil {
			return nil, err
		}

		chapters := []types.Chapter{}

		for _, ch := range data.Comic.Chapters {
			date, err := time.Parse(time.RFC3339Nano, ch.PublishedOn)
			if err != nil {
				return nil, err
			}

			chapters = append(chapters, types.Chapter{
				Title: ch.FullTitle,
				Id:    ch.Url,
				Date:  date,
			})
		}

		return &types.Manga{
			Synopsis:  data.Comic.Description,
			Author:    data.Comic.Author,
			Artist:    data.Comic.Artist,
			Img:       data.Comic.Thumbnail,
			Chapters:  chapters,
			Sourceurl: "https://tuttoanimemanga.net" + id,
		}, nil
	},

	Chapter: func(manga, id string) ([]string, error) {
		var data chapterRes

		println(baseurl + "api" + id)

		err, _ := utils.GetAndJsonParse(baseurl+"api"+id, &data)
		if err != nil {
			return nil, err
		}

		return data.Chapter.Pages, nil
	},
}
